src/vm_cpu.rs

```diff
@@ -96,6 +100,246 @@ pub struct VmCpuTrapState {
+            {
+                break None;
+            }
+            let result = self.counter_index + self.mask_index;
+            let mask_bit_set = self.counter_mask & (1 << self.mask_index);
+            self.counter_mask &= !(1 << self.mask_index);
+            self.mask_index += 1;
+            if mask_bit_set != 0 {
+                break Some(result as usize);
+            }
+        }
+    }
+}
+
+impl VmPmuState {
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to configure counters.
+    pub fn update_configured_counter(
+        &mut self,
+        platform_counter_index: PlatformCounterIndex,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<()> {
+        use PmuCounterState::*;
+        let new_state = CounterState {
+            config_flags,
+            event_type,
+            event_data,
+            value: 0,
+        };
+        let pmu_info = pmu::PmuInfo::get()?;
+        let counter_index = pmu_info.get_logical_counter_index(platform_counter_index)?;
+        let state = &mut self.counter_state[counter_index.0 as usize];
+        match state {
+            // If skip_match is set, counter must already be configured
+            Configured(c) | Started(c) if config_flags.get_skip_match() => {
+                if config_flags.get_auto_start() {
+                    *state = Started(*c);
+                }
+                Ok(())
+            }
+            NotConfigured if !config_flags.get_skip_match() => {
+                *state = if config_flags.get_auto_start() {
+                    Started(new_state)
+                } else {
+                    Configured(new_state)
+                };
+                Ok(())
+            }
+            _ => Err(SbiError::InvalidParam),
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to start counters.
+    pub fn update_started_counters(
+        &mut self,
+        counter_index: LogicalCounterIndex,
+        counter_mask: LogicalCounterMask,
+    ) {
+        use PmuCounterState::*;
+        // Unwrap ok: counter_index and counter_mask were previously validated
+        let bitmask_iter = CounterMaskIter::new(counter_index.0, counter_mask.0).unwrap();
+        for i in bitmask_iter {
+            let state = &mut self.counter_state[i];
+            match state {
+                // The Started(_) handles the corner case where some counters are automatically
+                // started on configuration, and the subsequent call to start_counters
+                // fails with "already started" (example: insret).
+                Started(c) | Configured(c) => {
+                    *state = Started(*c);
+                }
+                _ => unreachable!(),
```

Comment:
Hm, because you're using `any()` to check if there's a startable counter in a mask and the OpenSBI implementation will skip over counters that it can't start without returning an error, I don't think this case is actually unreachable. Similar deal for `update_stopped_counters()`.